engine/orchestration/src/main/java/org/apache/cloudstack/engine/orchestration/NetworkOrchestrator.java

```diff
@@ -883,6 +885,73 @@ public void saveExtraDhcpOptions(final String networkUuid, final Long nicId, fin
+        String requestedIpv4Address = requestedNicProfile.getRequestedIPv4();
+        if (requestedIpv4Address == null) {
+            s_logger.debug(String.format("The requested [IPv4 address=%s] is null", requestedIpv4Address));
+            return;
+        }
+        if (org.apache.commons.lang3.StringUtils.isBlank(requestedIpv4Address)) {
+            throw new InvalidParameterValueException(String.format("The requested [IPv4 address='%s'] is empty or blank", requestedIpv4Address));
+        }
+        if (!NetUtils.isValidIp4(requestedIpv4Address)) {
+            throw new InvalidParameterValueException(String.format("The requested [IPv4 address='%s'] is not a valid IP address", requestedIpv4Address));
+        }
+
+        VlanVO vlanVo = _vlanDao.findByNetworkId(network.getId());
+        if (vlanVo == null) {
+            throw new CloudRuntimeException(String.format("Trying to configure a Nic with the requested [IPv4='%s'] but cannot find a Vlan for the [network id='%s']",
+                    requestedIpv4Address, network.getId()));
+        }
+
+        acquireLockAndCheckIfIpv4IsFree(network, requestedIpv4Address);
+
+        String ipv4Gateway = vlanVo.getVlanGateway();
+        String ipv4Netmask = vlanVo.getVlanNetmask();
+
+        if (org.apache.commons.lang3.StringUtils.isBlank(ipv4Gateway) || !NetUtils.isValidIp4(ipv4Gateway)) {
+            throw new CloudRuntimeException(String.format("The [IPv4Gateway='%s'] from [VlanId='%s'] is not valid", ipv4Gateway, vlanVo.getId()));
+        }
+        if (org.apache.commons.lang3.StringUtils.isBlank(ipv4Netmask) || !NetUtils.isValidIp4Netmask(ipv4Netmask)) {
+            throw new CloudRuntimeException(String.format("The [IPv4Netmask='%s'] from [VlanId='%s'] is not valid", ipv4Netmask, vlanVo.getId()));
+        }
+
+        nicProfile.setIPv4Address(requestedIpv4Address);
+        nicProfile.setIPv4Gateway(ipv4Gateway);
+        nicProfile.setIPv4Netmask(ipv4Netmask);
+
+        if (nicProfile.getMacAddress() == null) {
+            try {
+                String macAddress = _networkModel.getNextAvailableMacAddressInNetwork(network.getId());
+                nicProfile.setMacAddress(macAddress);
+            } catch (InsufficientAddressCapacityException e) {
+                e.printStackTrace();
+            }
+        }
+    }
+
+    /**
+     *  Acquires lock of in table "user_ip_address" and checks if the requested IPv4 address is Free.
+     */
+    protected void acquireLockAndCheckIfIpv4IsFree(Network network, String requestedIpv4Address) {
+        IPAddressVO ipVO = _ipAddressDao.findByIpAndSourceNetworkId(network.getId(), requestedIpv4Address);
+        if (ipVO == null) {
+            throw new CloudRuntimeException(String.format("Cannot find IPAddressVO for guest [IPv4 address='%s'] and [network id=%s]", requestedIpv4Address, network.getId()));
+        }
+        IPAddressVO lockedIpVO = _ipAddressDao.acquireInLockTable(ipVO.getId());
```

Comment:
`lockedIpVO` can be `null` (https://github.com/apache/cloudstack/blob/master/framework/db/src/main/java/com/cloud/utils/db/GenericDaoBase.java#L1049). Moreover, it would be better IMO to put the lock/release operation with a `try/finally` block to ensure to release the lock all the time (when it could get it).

Reply:
@marcaurele are you LGTM on this one?
Thanks for your reviews and hard work :+1: